hydra/_internal/utils.py

```diff
@@ -525,32 +542,60 @@ def _getcls(field: str) -> str:
         assert isinstance(classname, str)
         return classname
 
-    def _has_field(field: str) -> bool:
-        if isinstance(config, DictConfig):
-            if field in config:
-                ret = config[field] != "???"
-                assert isinstance(ret, bool)
-                return ret
-            else:
-                return False
-        else:
-            if hasattr(config, field):
-                ret = getattr(config, field) != "???"
-                assert isinstance(ret, bool)
-                return ret
-            else:
-                return False
-
-    if _has_field(field="class"):
+    if _has_field(config=config, field="class"):
         _warn(field="class")
-    elif _has_field(field="cls"):
+    elif _has_field(config=config, field="cls"):
         _warn(field="cls")
+    elif _has_field(config=config, field="target"):
+        _warn(field="target")
 
-    if _has_field(field="target"):
+    if _has_field(config=config, field="type"):
+        return _getcls(field="type")
+    elif _has_field(config=config, field="target"):
         return _getcls(field="target")
-    elif _has_field(field="cls"):
+    elif _has_field(config=config, field="cls"):
         return _getcls(field="cls")
-    elif _has_field(field="class"):
+    elif _has_field(config=config, field="class"):
         return _getcls(field="class")
     else:
-        raise ValueError("Input config does not have a target field")
+        raise ValueError("Input config does not have a type field")
+
+
+def _get_callable_name(config: Union[ObjectConf, DictConfig]) -> str:
```

Comment:
this function is a copy of the other one with some minor changes. try to reduce this duplication by passing additional parameters. to one function.

Reply:
Okay!